core/unit_test/TestLegionInteroperability.cpp

```diff
@@ -0,0 +1,139 @@
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+
+// The purpose of this compile-only test is twofold:
+// 1. mimic Legion's use of Kokkos implementation details for initializing the
+//    exectution environment
+// 2. demonstrate how to leverage SFINAE to support Kokkos version through the
+//    ExecutionSpace::impl_initialize breaking change before release 3.7
+namespace {
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, "")  // FIXME C++17
+
+#ifdef KOKKOS_ENABLE_CUDA
+template <class T>
+using deprecated_cuda_impl_initialize_t =
+    decltype(T::impl_initialize(typename T::SelectDevice(0), 1));
+
+template <class T>
+constexpr bool has_deprecated_cuda_impl_initialize_v =
+    Kokkos::is_detected<deprecated_cuda_impl_initialize_t, T>::value;
+
+template <class T>
+std::enable_if_t<has_deprecated_cuda_impl_initialize_v<T>>
+legion_initilialize_kokkos_cuda() {
+  int cuda_device_id = 0;
+  int num_instances  = 1;
+  T::impl_initialize(typename T::SelectDevice(cuda_device_id), num_instances);
+}
```

Comment:
This will never be executed, right? Could we just wrap it with a `KOKKOS_ENABLE_DEPRECATED_CODE_3` guard so that we only keep it until the 4.0 release.?

Reply:
Never executed no.
Why would we not keep it longer?

Reply:
I don't like keeping code that is never executed.

Reply:
It is a compile only unit test.   We don't ever execute any of it.

Reply:
If you are requesting changes (not clear to me) what would you like this test to do?  Or are you arguing against having that test in the first place?

Reply:
If this overload is never chosen, I wouldn't introduce it. 
I think it's valuable to have a test for the Legion use case that actually executes the initialization of a specific backend and not just tries to compile the code, though. AFAIK, we don't have any other test that does this.

Reply:
I do not want to add a separate executable that we actually run.  What would we even do in that test?  What if neither Serial, OpenMP, nor Cuda is enabled?
If you feel strongly about this or others agree with you, I'd rather withdraw that change.

Reply:
I would add an executable that tries initializing all available backends individually as a follow-up since I find it helpful to ensure we don't break that capability. I would be curious to hear if that would find appreciation by others or not.

Either way, I don't think it's worth holding this pull request over this.